programs/clearing_house/src/controller/position.rs

```diff
@@ -66,43 +70,61 @@ pub struct PositionDelta {
     pub base_asset_amount: i128,
 }
 
-enum PositionUpdateType {
-    Open,
-    Increase,
-    Reduce,
-    Close,
-    Flip,
-}
-
-fn get_position_update_type(
-    position: &MarketPosition,
+pub fn update_amm_position(
+    market: &mut Market,
     delta: &PositionDelta,
-) -> PositionUpdateType {
-    if position.base_asset_amount == 0 {
-        PositionUpdateType::Open
-    } else if position.base_asset_amount.signum() == delta.base_asset_amount.signum() {
-        PositionUpdateType::Increase
-    } else if position.base_asset_amount.abs() > delta.base_asset_amount.abs() {
-        PositionUpdateType::Reduce
-    } else if position.base_asset_amount.abs() == delta.base_asset_amount.abs() {
-        PositionUpdateType::Close
+    is_per_lp_position: bool,
+) -> ClearingHouseResult<i128> {
+    let mut position = if is_per_lp_position {
+        market.amm.market_position_per_lp
+    } else {
+        market.amm.market_position
+    };
+
+    let new_position = position.base_asset_amount == 0;
+    let (new_quote_asset_amount, new_quote_entry_amount, new_base_asset_amount, pnl) =
+        calculate_position_new_quote_base_pnl(&position, delta)?;
+
+    let closed_position = new_base_asset_amount == 0;
+    let flipped_position = position.base_asset_amount.signum() != new_base_asset_amount.signum();
+
+    // Update user position
+    if closed_position {
+        position.last_cumulative_funding_rate = 0;
+        position.last_funding_rate_ts = 0;
+    } else if new_position || flipped_position {
+        if new_base_asset_amount > 0 {
+            position.last_cumulative_funding_rate = market.amm.cumulative_funding_rate_long;
+        } else {
+            position.last_cumulative_funding_rate = market.amm.cumulative_funding_rate_short;
+        }
+    }
+
+    position.quote_asset_amount = new_quote_asset_amount;
+    position.quote_entry_amount = new_quote_entry_amount;
+    position.base_asset_amount = new_base_asset_amount;
+
+    if is_per_lp_position {
+        market.amm.market_position_per_lp = position;
     } else {
-        PositionUpdateType::Flip
+        market.amm.market_position = position;
     }
+
+    Ok(pnl)
 }
 
 pub fn update_position_and_market(
     position: &mut MarketPosition,
     market: &mut Market,
     delta: &PositionDelta,
 ) -> ClearingHouseResult<i128> {
-    validate!(
-        delta.base_asset_amount != 0 && delta.quote_asset_amount != 0,
-        ErrorCode::InvalidPositionDelta,
-        "delta.base_asset_amount {} delta.quote_asset_amount {}",
-        delta.base_asset_amount,
-        delta.quote_asset_amount,
-    )?;
+    // validate!(
```

Comment:
why was this commented out? @0xbigz

Reply:
tried to write it so it can support updating position with only quote asset amount as can occur in `settle_lp_position`